numpy/lib/index_tricks.py

```diff
@@ -596,6 +597,419 @@ def __iter__(self):
     next = __next__
 
 
+class ndrange(_nx.collections_abc.Sequence):
+    """
+    The ndrange type represents an immutable sequence of numbers and can be
+    used for looping over specific multi-dimensional indicies in ``for`` loops.
+
+    # TODO: how do I get this to show up in the docs
+    class ndrange(stop)
+    class ndrange(start, stop[, step])
+
+    The arguments to the ``ndrange`` constructor must be iterables (lists,
+    tuple, 1D-arrays, etc.) of integers. Individual integers are also accepted
+    and are coerced to tuples of length 1.
+
+    The length of all provided arguments must be equal.
+
+    While conceptually similar to a multidimensional object, the iterator
+    returned by ndrange is a flat iterator that would be determined by the
+    array of tuples
+
+      r[i_0, i_1, ..., i_Nminus1] = (start[0] + step[0] * i_0,
+                                     start[1] + step[1] * i_1,
+                                     ...,
+                                     start[N-1] + step[N-1] * i_Nminus1)
+
+    where ``i0, i1, ..., iN_m1 >= 0``. For positive steps, at an axis ``j``,
+    the largest index is constrained such that
+    ``r[..., i_j, ...][j] < stop[j]``. For negative steps at an axis ``j``,
+    sthe smallest index at that axis is contrained by
+    ``r[..., i_j, ...][j] > stop[j]``.
+
+    An ndrange object will be empty if `r[(0,) * len(start)]` does not meet
+    the value constraints.
+
+    The sequence of tuples is traversed the conceptual matrix ``r`` in a
+    fashion consistent with C order indexing.
+
+    While the ndrange range returns a flat sequence, it can be sliced along
+    multiple axes. As an added convenience, the ``shape`` attribute is also
+    available, and returns the conceptual shape of the array ``r`` associated
+    which the sequence traverses.
```

Comment:
To be more precise, ndrange does not "return" a flat sequence, it iterates as a flat sequence. But that sentence clause is unneeded, you can just start off with "The ndrange object can be sliced along multiple axes, and has a ``shape`` attribute corresponding to the number of indexed elements along each dimension".

Reply:
See new wording.